pyccel/ast/core.py

```diff
@@ -1572,7 +1572,8 @@ def __init__(self, name, *, value = None, kwonly=False, annotation=None, bound_a
             name.declare_as_argument()
 
         if pyccel_stage != "syntactic":
-            self._inout = self.var.rank>0 and not self.var.is_const if isinstance(self.var, Variable) else False
+            self._inout = (self.var.rank > 0 or isinstance(self.var.class_type, CustomDataType)) \
+                        and not self.var.is_const if isinstance(self.var, Variable) else False
```

Comment:
```suggestion
            self._inout = (self.var.rank > 0 or isinstance(self.var.class_type, CustomDataType)) 
            self._inout &= not (isinstance(self.var, Variable) and self.var.is_const)
```

Reply:
This correction was incorrect. The "if" originally had precedence over the "and" so the result is False if the var is not a Variable (i.e. if it is a FunctionAddress)